Panic when updateCqStatusIfChanged in function of Pending() in cluster_queue

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:
Panic when updating cluster queue status, error logs below
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa0 pc=0x195926c]

goroutine 1014 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x2402060?, 0x420b580?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
sigs.k8s.io/kueue/pkg/queue.(*ClusterQueue).Pending(0x23d8420?)
	/workspace/pkg/queue/cluster_queue.go:283 +0x2c
sigs.k8s.io/kueue/pkg/queue.(*Manager).Pending(0xc064e6b788?, 0xc047b0bdc0?)
	/workspace/pkg/queue/manager.go:279 +0xa5
sigs.k8s.io/kueue/pkg/controller/core.(*ClusterQueueReconciler).updateCqStatusIfChanged(0xc064e12fa0, {0x2c2dda8, 0xc0671cb050}, 0xc064e6b600, {0x2804ec5, 0x5}, {0x2809bc0, 0x8}, {0x2824f6c, 0x16})
	/workspace/pkg/controller/core/clusterqueue_controller.go:647 +0xa8
sigs.k8s.io/kueue/pkg/controller/core.(*ClusterQueueReconciler).Reconcile(0xc064e12fa0, {0x2c2dda8, 0xc0671cae40}, {{{0x0?, 0xa32964?}, {0xc00100f530?, 0x41ac18?}}})
	/workspace/pkg/controller/core/clusterqueue_controller.go:198 +0x5eb
sigs.k8s.io/kueue/pkg/controller/core.(*leaderAwareReconciler).Reconcile(0xc04f718a80, {0x2c2dda8, 0xc0671cae40}, {{{0x0?, 0x5?}, {0xc00100f530?, 0xc065035d10?}}})
	/workspace/pkg/controller/core/leader_aware_reconciler.go:77 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2c363b8?, {0x2c2dda8?, 0xc0671cae40?}, {{{0x0?, 0xb?}, {0xc00100f530?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc064e13040, {0x2c2dde0, 0xc00069eb40}, {0x253e3a0, 0xc06520a360})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:316 +0x3bc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc064e13040, {0x2c2dde0, 0xc00069eb40})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266 +0x1be
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 820
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:223 +0x50c

**How to reproduce it (as minimally and precisely as possible)**:

Not sure how to easily repro it yet. this is observed when we do the capacity updating for a cohort
 
**Anything else we need to know?**:
1. check further log, it saying the panic happens when updating the status for one cluster-queue, let’s say cq-a, there is only log related to this cq-a listed as below

{"level":"Level(-2)","ts":"2024-06-20T16:57:50.608225383Z","caller":"core/clusterqueue_controller.go:167","msg":"Reconciling ClusterQueue","controller":"clusterqueue","controllerGroup":"kueue.x-k8s.io",

**Environment**:
- Kubernetes version (use `kubectl version`): v1.28.9-eks-036c24b
- Kueue version (use `git describe --tags --dirty --always`): 0.7
- Cloud provider or hardware configuration: aws
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/cache/cache.go
+++ b/pkg/cache/cache.go
@@ -39,7 +39,7 @@ import (
 )
 
 var (
-	errCqNotFound          = errors.New("cluster queue not found")
+	ErrCqNotFound          = errors.New("cluster queue not found")
 	errQNotFound           = errors.New("queue not found")
 	errWorkloadNotAdmitted = errors.New("workload not admitted by a ClusterQueue")
 )
@@ -369,7 +369,7 @@ func (c *Cache) UpdateClusterQueue(cq *kueue.ClusterQueue) error {
 	defer c.Unlock()
 	cqImpl, ok := c.clusterQueues[cq.Name]
 	if !ok {
-		return errCqNotFound
+		return ErrCqNotFound
 	}
 	if err := cqImpl.update(cq, c.resourceFlavors, c.admissionChecks); err != nil {
 		return err
@@ -503,7 +503,7 @@ func (c *Cache) DeleteWorkload(w *kueue.Workload) error {
 
 	cq := c.clusterQueueForWorkload(w)
 	if cq == nil {
-		return errCqNotFound
+		return ErrCqNotFound
 	}
 
 	c.cleanupAssumedState(w)
@@ -547,7 +547,7 @@ func (c *Cache) AssumeWorkload(w *kueue.Workload) error {
 
 	cq, ok := c.clusterQueues[string(w.Status.Admission.ClusterQueue)]
 	if !ok {
-		return errCqNotFound
+		return ErrCqNotFound
 	}
 
 	if err := cq.addWorkload(w); err != nil {
@@ -572,7 +572,7 @@ func (c *Cache) ForgetWorkload(w *kueue.Workload) error {
 
 	cq, ok := c.clusterQueues[string(w.Status.Admission.ClusterQueue)]
 	if !ok {
-		return errCqNotFound
+		return ErrCqNotFound
 	}
 	cq.deleteWorkload(w)
 	if c.podsReadyTracking {
@@ -596,7 +596,7 @@ func (c *Cache) Usage(cqObj *kueue.ClusterQueue) (*ClusterQueueUsageStats, error
 
 	cq := c.clusterQueues[cqObj.Name]
 	if cq == nil {
-		return nil, errCqNotFound
+		return nil, ErrCqNotFound
 	}
 
 	stats := &ClusterQueueUsageStats{

--- a/pkg/controller/core/clusterqueue_controller.go
+++ b/pkg/controller/core/clusterqueue_controller.go
@@ -644,7 +644,11 @@ func (r *ClusterQueueReconciler) updateCqStatusIfChanged(
 	reason, msg string,
 ) error {
 	oldStatus := cq.Status.DeepCopy()
-	pendingWorkloads := r.qManager.Pending(cq)
+	pendingWorkloads, err := r.qManager.Pending(cq)
+	if err != nil {
+		r.log.Error(err, "Failed getting pending workloads from queue manager")
+		return err
+	}
 	stats, err := r.cache.Usage(cq)
 	if err != nil {
 		r.log.Error(err, "Failed getting usage from cache")

--- a/pkg/controller/core/clusterqueue_controller_test.go
+++ b/pkg/controller/core/clusterqueue_controller_test.go
@@ -49,18 +49,23 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 	}
 
 	testCases := map[string]struct {
-		cqStatus           kueue.ClusterQueueStatus
-		newConditionStatus metav1.ConditionStatus
-		newReason          string
-		newMessage         string
-		newWl              *kueue.Workload
-		wantCqStatus       kueue.ClusterQueueStatus
+		insertCqIntoCache   bool
+		insertCqIntoManager bool
+		cqStatus            kueue.ClusterQueueStatus
+		newConditionStatus  metav1.ConditionStatus
+		newReason           string
+		newMessage          string
+		newWl               *kueue.Workload
+		wantCqStatus        kueue.ClusterQueueStatus
+		wantError           error
 	}{
 		"empty ClusterQueueStatus": {
-			cqStatus:           kueue.ClusterQueueStatus{},
-			newConditionStatus: metav1.ConditionFalse,
-			newReason:          "FlavorNotFound",
-			newMessage:         "Can't admit new workloads; some flavors are not found",
+			insertCqIntoCache:   true,
+			insertCqIntoManager: true,
+			cqStatus:            kueue.ClusterQueueStatus{},
+			newConditionStatus:  metav1.ConditionFalse,
+			newReason:           "FlavorNotFound",
+			newMessage:          "Can't admit new workloads; some flavors are not found",
 			wantCqStatus: kueue.ClusterQueueStatus{
 				PendingWorkloads: int32(len(defaultWls.Items)),
 				Conditions: []metav1.Condition{{
@@ -73,6 +78,8 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 			},
 		},
 		"same condition status": {
+			insertCqIntoCache:   true,
+			insertCqIntoManager: true,
 			cqStatus: kueue.ClusterQueueStatus{
 				PendingWorkloads: int32(len(defaultWls.Items)),
 				Conditions: []metav1.Condition{{
@@ -97,6 +104,8 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 			},
 		},
 		"same condition status with different reason and message": {
+			insertCqIntoCache:   true,
+			insertCqIntoManager: true,
 			cqStatus: kueue.ClusterQueueStatus{
 				PendingWorkloads: int32(len(defaultWls.Items)),
 				Conditions: []metav1.Condition{{
@@ -121,6 +130,8 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 			},
 		},
 		"different condition status": {
+			insertCqIntoCache:   true,
+			insertCqIntoManager: true,
 			cqStatus: kueue.ClusterQueueStatus{
 				PendingWorkloads: int32(len(defaultWls.Items)),
 				Conditions: []metav1.Condition{{
@@ -145,6 +156,8 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 			},
 		},
 		"different pendingWorkloads with same condition status": {
+			insertCqIntoCache:   true,
+			insertCqIntoManager: true,
 			cqStatus: kueue.ClusterQueueStatus{
 				PendingWorkloads: int32(len(defaultWls.Items)),
 				Conditions: []metav1.Condition{{
@@ -169,6 +182,13 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 				}},
 			},
 		},
+		"cluster queue does not exist on manager": {
+			wantError: queue.ErrClusterQueueDoesNotExist,
+		},
+		"cluster queue does not exist on cache": {
+			insertCqIntoManager: true,
+			wantError:           cache.ErrCqNotFound,
+		},
 	}
 
 	for name, tc := range testCases {
@@ -186,11 +206,15 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 				Build()
 			cqCache := cache.New(cl)
 			qManager := queue.NewManager(cl, cqCache)
-			if err := cqCache.AddClusterQueue(ctx, cq); err != nil {
-				t.Fatalf("Inserting clusterQueue in cache: %v", err)
+			if tc.insertCqIntoCache {
+				if err := cqCache.AddClusterQueue(ctx, cq); err != nil {
+					t.Fatalf("Inserting clusterQueue in cache: %v", err)
+				}
 			}
-			if err := qManager.AddClusterQueue(ctx, cq); err != nil {
-				t.Fatalf("Inserting clusterQueue in manager: %v", err)
+			if tc.insertCqIntoManager {
+				if err := qManager.AddClusterQueue(ctx, cq); err != nil {
+					t.Fatalf("Inserting clusterQueue in manager: %v", err)
+				}
 			}
 			if err := qManager.AddLocalQueue(ctx, lq); err != nil {
 				t.Fatalf("Inserting localQueue in manager: %v", err)
@@ -207,9 +231,9 @@ func TestUpdateCqStatusIfChanged(t *testing.T) {
 			if tc.newWl != nil {
 				r.qManager.AddOrUpdateWorkload(tc.newWl)
 			}
-			err := r.updateCqStatusIfChanged(ctx, cq, tc.newConditionStatus, tc.newReason, tc.newMessage)
-			if err != nil {
-				t.Errorf("Updating ClusterQueueStatus: %v", err)
+			gotError := r.updateCqStatusIfChanged(ctx, cq, tc.newConditionStatus, tc.newReason, tc.newMessage)
+			if diff := cmp.Diff(tc.wantError, gotError, cmpopts.EquateErrors()); len(diff) != 0 {
+				t.Errorf("Unexpected error (-want/+got):\n%s", diff)
 			}
 			configCmpOpts := []cmp.Option{
 				cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"),

--- a/pkg/queue/manager.go
+++ b/pkg/queue/manager.go
@@ -37,7 +37,7 @@ import (
 
 var (
 	ErrQueueDoesNotExist         = errors.New("queue doesn't exist")
-	errClusterQueueDoesNotExist  = errors.New("clusterQueue doesn't exist")
+	ErrClusterQueueDoesNotExist  = errors.New("clusterQueue doesn't exist")
 	errClusterQueueAlreadyExists = errors.New("clusterQueue already exists")
 )
 
@@ -158,7 +158,7 @@ func (m *Manager) UpdateClusterQueue(ctx context.Context, cq *kueue.ClusterQueue
 	defer m.Unlock()
 	cqImpl, ok := m.clusterQueues[cq.Name]
 	if !ok {
-		return errClusterQueueDoesNotExist
+		return ErrClusterQueueDoesNotExist
 	}
 
 	oldCohort := cqImpl.Cohort()
@@ -273,10 +273,16 @@ func (m *Manager) PendingWorkloads(q *kueue.LocalQueue) (int32, error) {
 	return int32(len(qImpl.items)), nil
 }
 
-func (m *Manager) Pending(cq *kueue.ClusterQueue) int {
+func (m *Manager) Pending(cq *kueue.ClusterQueue) (int, error) {
 	m.RLock()
 	defer m.RUnlock()
-	return m.clusterQueues[cq.Name].Pending()
+
+	cqImpl, ok := m.clusterQueues[cq.Name]
+	if !ok {
+		return 0, ErrClusterQueueDoesNotExist
+	}
+
+	return cqImpl.Pending(), nil
 }
 
 func (m *Manager) QueueForWorkloadExists(wl *kueue.Workload) bool {

